### ISSUE TITLE
CloudFront with Web Application Firewall?

### ISSUE THREAD

**Issue by blazinaj** [aws-amplify/amplify-console#8]
** Which Category is your question related to? **
Amplify Console Hosting

If I'm using the Amplify Console to host a Static Website, where is it stored? It doesn't look like there is an S3 bucket being used by the Amplify Console to host the website?

If we use the Amplify CLI `add hosting`, it will use an S3 bucket, does the Amplify Console do something similar, but we just can't access it?

**Comment by garyleefight** [comment]
@blazinaj Hi, you can chain your customize CloudFront distribution in front of the managed CloudFront distribution. You have to pass '_DISABLE_L2_CACHE: true' environment variable to CodeBuild to disable service CloudFront chain.

**Comment by blazinaj** [comment]
Thank you very much. That is what I needed. What does the `_DISABLE_l2_CACHE: true` do exactly?

**Comment by blazinaj** [comment]
When I'm creating a CloudFront distribution, do I put the full amplify url as the *Origin Domain Name Name*?

![image](https://user-images.githubusercontent.com/39976117/62558969-ea466880-b82e-11e9-90d6-d3249c1541dd.png)

**Comment by garyleefight** [comment]
> Thank you very much. That is what I needed. What does the `_DISABLE_l2_CACHE: true` do exactly?

Disable the L2 cache will disable one layer of CloudFront that AWS Amplify Console used for instant deployment. As CloudFront allows 2 maximum CloudFront chaining together, you have to disable it and add your customize distribution.

Yes. You have to input the full domain name points to AWS Amplify domain

**Comment by blazinaj** [comment]
Thank you! 

I got it hooked up through my custom CloudFront distribution. How do I make it so that the `production.abcxyz.amplifyapp.com` URL doesn't just bypass my custom CloudFront/WAF?

**Comment by garyleefight** [comment]
@blazinaj Hi, for the WAF feature the original URL is still public. The workaround can't disable the original domain. Right now, you can only use basic auth to ensure its security.

**Comment by blazinaj** [comment]
Ah, thank you very much. That's what I needed to know. I'll submit a feature request for that.

**Comment by dave-moser** [comment]
If we are also using a custom domain with our Console app, should this CloudFront/WAF workaround work?  If so, what would I use for the Origin in the CloudFront Distribution - our custom domain name or the Console generated domain?  I tried our custom domain and traffic did not seem to get routed properly through Cloudfront to get to the Console app.  It seems to just go straight to the Console app.

**Comment by dpalmeira** [comment]
Hi,
I have the same situation of previous message (dave-moser), custom domain, and I want to protect it with WAF and IpFiltering with IPSet. Could somebody answer if the @garyleefight workaround and the configuration indicated by @blazinaj should work with custom domains added in the Amplify Console?

thank you

**Comment by kraichura** [comment]
If anyone comes across this later, what we have done is to have custom cloudfront with a customized WAF. We added a custom name to the cloudfronts cname. The original domain name of Amplify is not customized (hence the customization of the domain name happens in cloudfront). We added access control to Amplify, so anyone with the original public URL cannot access amplify directly. We then added an authorization header as the custom header to cloudfront origin, so that we can access the amplify app s3 static pages from the cloudfront domain.

**Comment by BrokenCasio** [comment]
@kraichura I am trying to take your approach my amplify cloudfront is protected via the amplify restricted access, I have created a second cloudfront with the necessary WAF rules however I am having trouble bridging it to the amplify cloudfront. My authorization headers are not bypassing the restriction, can you share the Origin Custom Headers you used and if any special cache or origin request policies came into play ? Thanks.

**Comment by qbig** [comment]
@kraichura @BrokenCasio @blazinaj  Do you have to set environment var "_DISABLE_l2_CACHE: true" for your branch in order for the additional cloudfront WAF to work ? 
I am getting "err_too_many_redirects" for my custom cloudfront distribution and it doesn't redirect to the origin

**Comment by qbig** [comment]
apparently, we need to 1. set the environment var; then 2. create cloudfront distribution; the sequence seems matters as my first attempt didn't work

**Comment by blazinaj** [comment]
@qbig I was never able to get this working, we moved away from Amplify Console and are using the 'Hosting' category directly through the CLI. This deploys to an S3 bucket and then we manually set up and managed WAF with CloudFront

**Comment by kriscodes2** [comment]
I adopted Amplify in December of 2018 just after it was introduced at re-Invent.  In the beginning "**_it was_**" possible to directly manipulate the CloudFront distribution and S3 deployment bucket. In my humble opinion Amplify was the answer to the microservices architecture that was booming since 2011. Shortly after Kubernetes came about in June 2014, Serverless came on the scene after AWS Lambda's debut in 2014 and out of beta in October 2015 .  All are just abstractions from its predecessors. 

To the point..... I believe that control is purposely being abstracted away in lieu of easy drag and drop options. To solve this as developers I feel we need to learn more and sometimes go backwards. I have used Amplify heavily since its debut and had to learn many other technologies to implement custom features like VPC's, CloudFront templates, Lambda and SQL/NoSQL. Try these other services to build your own solutions or maybe give the serverless Framework a try...
